fairlearn/reductions/_moments/conditional_selection_rate.py

```diff
@@ -15,27 +15,45 @@
+    This serves as the base class for :class:`DemographicParity` :class:`EqualizedOdds`, and
+    others. All subclasses can be used as difference-based constraints or ratio-based constraints.
+
+    Difference-based constraints are written with a difference bound :math:`\\epsilon`.
+    Ratio-based constraints have a similar :math:`\\epsilon`, referred to as ratio bound slack
+    below. Additionally, they require a ratio bound. Constraints are set up for each group
+    compared to the overall population (unless further events are specified, e.g., in Equalized
+    Odds). Constraint violation for difference-based constraints starts if the difference between
+    a group and the overall population with regard to a utility exceeds the difference bound. For
+    ratio-based constraints we consider the difference between the overall population's utility
+    and a group's utility multiplied with the ratio bound.
 
     The `index` field is a :class:`pandas:pandas.MultiIndex` corresponding to the rows of
-    the DataFrames either required as arguments or returned by several
-    of the methods of the `ConditionalSelectionRate` class. It is the cartesian
-    product of:
+    the DataFrames either required as arguments or returned by several of the methods of the
+    `ConditionalSelectionRate` class. It is the cartesian product of:
 
     - The unique events defined for the particular object
     - The unique values for the sensitive feature
     - The characters `+` and `-`, corresponding to the Lagrange multipliers
       for positive and negative violations of the constraint
 
-    The `ratio` specifies the multiple at which error(A = a) should be compared with total_error
-    and vice versa. The value of `ratio` has to be in the range (0,1] with smaller values
-    corresponding to weaker constraint. The `ratio` equal to 1 corresponds to the constraint
-    where error(A = a) = total_error
+    :param difference_bound: the constraints' difference bound for constraints that are expressed
+        as differences, often referred to as :math:`\\epsilon`.
+        If `ratio_bound` is used then `difference_bound` needs to be None.
+        If neither `ratio_bound` nor `difference_bound` are set then a default difference bound of
+        0.01 is used for backwards compatibility. Default None.
+    :type difference_bound: float
+    :param: ratio_bound: the constraints' ratio bound for constraints that are expressed as
+        ratios. The specified value needs to abide by :math:`0 < \\text{ratio_bound} \\leq 1`.
+        If `difference_bound` is used then `ratio_bound` needs to be None.
+        Default None.
+    :type ratio_bound: float
+    :param ratio_bound_slack: the constraints' ratio bound slack for constraints that are
+        expressed as ratios, usually referred to as :math:`\\epsilon`.
+        `ratio_bound_slack` is ignored if `ratio_bound` is not specified.
+        Default 0.0
+    :type ratio_bound_slack: float
```

Comment:
It seems like the decision to use a superclass here is obscuring how to use user-facing classes.  By using a common API across all subclasses, it means this documentation is all generic and describes a combination of options.  Further, it means there is no place in the code comments that directly express what the parameters to `DemographicParity` are, to pick one example.  Focusing on just the comments, what do you think about removing the comments here in the superclass, and instead updating the comments on the user-facing classes to describe to users how they can use the classes that the library provides, and how they can configure them?

Alternately, focusing on the code, a more direct code change might break the inheritance chain so that `DemographicParity` can stand alone, define its interface, and then internally use `ConditionalSelectionRate` to do whatever it needs, including parameterizing it properly and providing the common interface that's needed to the algorithms.  That would avoid the explosion of parameters in the superclass (and thus each subclass), and let user-facing classes focus on describing to users how they can use them.  I realize that's probably impossible since there are things in-flight right now, but otherwise it seems like anytime the team wants to change the parameters to any of the subclasses, it will require changing the superclass and also changing the parameters that all the other subclasses take.

Reply:
I'm writing a user guide that explains how to use these classes (the user facing ones). Please reevaluate the PR when that's there. Right now it's work in progress (WIP).

From a developer perspective these are perfectly reasonable things to do (using inheritance), and for people familiar with the code it's very useful to know what changed in this format. Your other comment indicates that you're looking for a user facing change description, and as I outlined there we should have that. That's not the purpose of this PR, though.

I don't think breaking inheritance makes sense since it's explicitly designed for this purpose. If you read Miro's proposal #9 in the proposals repo (it's actually accepted and checked in now) you may get the higher level perspective. All constraints have in common that we can express them as differences or ratios, that's why all of this can live together in the parent class. That means the API can stay like this. Again, the user guide section I'm writing will make this a lot clearer, so I'll have to ask for patience.

Reply:
@romanlutz oop!  I saw "widespread documentation updates for the webpage for reductions & moments in particular" in the pull request description at the time but missed the "WIP" in the title, sorry about that :)

@riedgar-ms It seems like the `mitigation` file in this pull request is what you were referring to in https://github.com/fairlearn/fairlearn/pull/480#discussion_r441066021, and the intention is for this PR to land first?

Either way I'm happy to help bring a helpful lack of experience and knowledge to reading and reviewing whenever that would actually be helpful :)

Reply:
@kevinrobinson Correct. I also bring a certain lack of experience to this area (I've had other priorities since the beginning of the year), which had been my motivation for starting to write the other document.